README.md
# planroom-ui [![Master Build Status](https://travis-ci.org/mjsmith11/planroom-ui.svg?branch=master)](https://travis-ci.org/mjsmith11/planroom-ui)

This repository contains a Vue.js front end for the planroom project.
It adheres to [Semantic Versioning](https://semver.org/) and [Gitflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow)

## Honorary Dependency
 - [lkf](https://github.cm/mjsmith11/lkf)

## Ubuntu Environment Setup
1. Setup an instance of [planroom-api](https://github.com/mjsmith11/planroom-api).
1. Ensure that the URL in .env.development is correct for your instance of planroom-api.
1. Install [npm](https://www.npmjs.com/get-npm).
1. In `public/index.html`, replace `<API_KEY HERE>` with an [api key](https://developers.google.com/maps/documentation/javascript/get-api-key) for the google maps and places apis.
1. Install dependencies `npm install`
1. Check to make sure the url for planroom api is correct in `.env.development`.
1. Run development server `npm run serve`

## Helpful Tools
 - Vue developer tools chrome add on.

### Install Dependencies
```
npm install
```

### Run Development Server
```
npm run serve
```

### Compiles and minifies for production
```
npm run build
```

### Lints and fixes files
```
npm run lint
```

### Run your unit tests
```
npm run test:unit
```



**For details on the contents of this repository, see [GETTING_STARTED](https://github.com/mjsmith11/planroom-ui/blob/master/GETTING_STARTED.md)**

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Changed
 - favicon to company branding
### Added
 - lkf honorary dependency
 - mailto link in job view
 - Handling of months in datetime filters
### Fixed
 - Taxable spelling

## [1.1.2] - 2018-11-17
### Fixed
 - Encode symbols correctly in upload filenames
 - Always download files instead of opening in the browser tab

## [1.1.1] - 2018-11-13
### Fixed
 - Datetime display for noon and midnight

## [1.1.0] - 2018-11-11
### Added
 - View for contractor to send invitations
 - View for subcontractor users
 - Routing for subcontractors
 - Vuex getters, action, and mutator for subcontractors

## [1.0.4] - 2018-10-28
### Changed
 - Make Dropzone use Planroom-Authorization header

## [1.0.3] - 2018-10-28
### Changed
 - Token header is now Planroom-Authorization

## [1.0.2] - 2018-10-28
### Fixed
 - Send API calls to https
### Added
  - .htaccess Rule to redirect http to https

## [1.0.0] - 2018-10-27
### Added
 - Navbar
 - Login component
 - Vuex store for authentication state
 - Route guards based on authentication status

## [0.3.0] - 2018-10-13
### Added
 - Uploading Files to a Job
 - Downloading Files from a Job
 
## [0.2.0] - 2018-09-27
### Fixed
 - New Job Form labels and field order
### Added
 - Job List View
 - Job Details View

## [0.1.0] - 2018-09-11
### Added 
 - CI Builds with Travis
 - Form for adding a job to the system
 
## [0.0.1] - 2018-08-20
### Added
 - Initial Empty Vue Project


tests/unit/App.spec.js
import { mount, createLocalVue } from '@vue/test-utils'
import App from '@/App'
import Vuex from 'vuex'
import axios from 'axios'

const localVue = createLocalVue()

localVue.use(Vuex)

describe('App', () => {
  it('has correct html structure logged in', () => {
    // This is a workaround because axios is still the mocked axios
    axios.interceptors = {}
    axios.interceptors.response = {}
    axios.interceptors.response.use = jest.fn()

    let getters = {
      isLoggedIn: () => true,
      isContractorUser: () => true
    }
    let store = new Vuex.Store({
      getters
    })
    const $route = {
      name: 'test'
    }
    let cmp = mount(App, {
      stubs: [
        'router-view',
        'router-link'
      ],
      store,
      localVue,
      mocks: {
        $route
      }

    })
    expect(cmp.html()).toMatchSnapshot()
  })
  it('has correct html structure logged out', () => {
    // This is a workaround because axios is still the mocked axios
    axios.interceptors = {}
    axios.interceptors.response = {}
    axios.interceptors.response.use = jest.fn()

    let getters = {
      isLoggedIn: () => false,
      isContractorUser: () => true
    }
    let store = new Vuex.Store({
      getters
    })
    let cmp = mount(App, {
      stubs: [
        'router-view',
        'router-link'
      ],
      store,
      localVue
    })
    expect(cmp.html()).toMatchSnapshot()
  })

  it('has correct html structure on jobs index route', () => {
    // This is a workaround because axios is still the mocked axios
    axios.interceptors = {}
    axios.interceptors.response = {}
    axios.interceptors.response.use = jest.fn()

    let getters = {
      isLoggedIn: () => true,
      isContractorUser: () => true
    }
    let store = new Vuex.Store({
      getters
    })
    const $route = {
      name: 'jobs index'
    }
    let cmp = mount(App, {
      stubs: [
        'router-view',
        'router-link'
      ],
      store,
      localVue,
      mocks: {
        $route
      }
    })
    expect(cmp.html()).toMatchSnapshot()
  })

  it('has correct html structure on jobs new job', () => {
    // This is a workaround because axios is still the mocked axios
    axios.interceptors = {}
    axios.interceptors.response = {}
    axios.interceptors.response.use = jest.fn()

    let getters = {
      isLoggedIn: () => true,
      isContractorUser: () => true
    }
    let store = new Vuex.Store({
      getters
    })
    const $route = {
      name: 'new job'
    }
    let cmp = mount(App, {
      stubs: [
        'router-view',
        'router-link'
      ],
      store,
      localVue,
      mocks: {
        $route
      }
    })
    expect(cmp.html()).toMatchSnapshot()
  })

  it('logs out', () => {
    // This is a workaround because axios is still the mocked axios
    axios.interceptors = {}
    axios.interceptors.response = {}
    axios.interceptors.response.use = jest.fn()

    let getters = {
      isLoggedIn: () => true,
      isContractorUser: () => true
    }
    let actions = {
      logout: jest.fn()
    }
    let store = new Vuex.Store({
      state: {},
      actions,
      getters
    })
    const $route = {
      name: 'jobs index'
    }
    const $router = {
      push: jest.fn()
    }
    let cmp = mount(App, {
      stubs: [
        'router-view',
        'router-link'
      ],
      store,
      localVue,
      mocks: {
        $route,
        $router
      }
    })
    cmp.find('#logout-link').trigger('click')
    expect(actions.logout).toHaveBeenCalled()
    expect($router.push).toHaveBeenCalledWith('/login')
  })
  it('renders for subcontractor', () => {
    // This is a workaround because axios is still the mocked axios
    axios.interceptors = {}
    axios.interceptors.response = {}
    axios.interceptors.response.use = jest.fn()

    let getters = {
      isLoggedIn: () => true,
      isContractorUser: () => false
    }
    let store = new Vuex.Store({
      getters
    })
    const $route = {
      name: 'test'
    }
    let cmp = mount(App, {
      stubs: [
        'router-view',
        'router-link'
      ],
      store,
      localVue,
      mocks: {
        $route
      }

    })
    expect(cmp.html()).toMatchSnapshot()
  })
})


tests/unit/Login.spec.js
import { mount } from '@vue/test-utils'
import Login from '@/views/Login'
import flushPromises from 'flush-promises'

describe('App', () => {
  it('has expected html', () => {
    let cmp = mount(Login, {

    })
    expect(cmp.html()).toMatchSnapshot()
  })

  it('handles successful login', async () => {
    const $router = {
      push: jest.fn()
    }
    const $store = {
      dispatch: jest.fn().mockReturnValue(new Promise((resolve, reject) => {
        resolve()
      }))
    }
    let cmp = mount(Login, {
      mocks: {
        $router,
        $store
      }
    })
    const emailField = cmp.find('#email')
    emailField.setValue('test@test.com')
    const passwordField = cmp.find('#password')
    passwordField.setValue('myPassword')

    cmp.find('form').trigger('submit')
    expect($store.dispatch).toBeCalledWith('login', {email: 'test@test.com', password: 'myPassword'})

    await flushPromises()

    expect($router.push).toHaveBeenCalledWith('/jobs')
  })

  it('handles failed login', async () => {
    const $router = {
      push: jest.fn()
    }
    const $store = {
      dispatch: jest.fn().mockReturnValue(new Promise((resolve, reject) => {
        reject(new Error('Login Failed'))
      }))
    }
    let cmp = mount(Login, {
      mocks: {
        $router,
        $store
      }
    })
    const emailField = cmp.find('#email')
    emailField.setValue('test@test.com')
    const passwordField = cmp.find('#password')
    passwordField.setValue('wrongPassword')

    cmp.find('form').trigger('submit')
    expect($store.dispatch).toBeCalledWith('login', { email: 'test@test.com', password: 'wrongPassword' })

    await flushPromises()

    expect($router.push).not.toHaveBeenCalled()
    const alert = cmp.find('div.alert')
    expect(alert.classes()).toContain('alert-danger')
    expect(alert.html()).toContain('Login Failed')
  })
})


tests/unit/store.spec.js
import store from '@/store'
import mockAxios from 'jest-mock-axios'
import axios from 'axios'
// import flushPromises from 'flush-promises'

// tokens can be decoded with https://jwt.io

describe('Vuex store', () => {
  beforeEach(() => {
    store.replaceState({
      status: '',
      token: '',
      user: '',
      refresher: ''
    })
  })
  it('gets status', () => {
    store.state.status = 'test status'
    expect(store.getters.authStatus).toBe('test status')
  })
  it('gets token', () => {
    store.state.token = 'test token'
    expect(store.getters.token).toBe('test token')
  })
  it('reports not logged in without a token', () => {
    expect(store.getters.isLoggedIn).toBe(false)
  })
  it('reports not logged in with expired token', () => {
    store.state.token = 'eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJleHAiOjEyMzQ1fQ.2QdEUMj0lBSB1276xa3yW6MNM5PdzKrUFYNSDKkD_NuB0kHnVV_4R8Rt1dISQ99f-Res-YSqnsCApbNudr1p3g'
    expect(store.getters.isLoggedIn).toBe(false)
  })
  it('reports logged in with valid token', () => {
    // make vuex think we are logged in.  Expires in 2361
    store.state.token = 'eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJleHAiOjEyMzQ1Njc3ODc3fQ.6xV-z88Nvmag8i4jVwmOZjX3MhCYAgb3rqttN4ROix3EbtHLwYIG3utNVaCpCN2cS7QFAJM3CPnfiS5_s9luiA'
    expect(store.getters.isLoggedIn).toBe(true)
  })
  it('gets job', () => {
    store.state.token = 'eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJleHAiOjEyMzQ1Njc3LCJyb2xlIjoic3ViY29udHJhY3RvciIsImpvYiI6IjciLCJlbWFpbCI6InRlc3RAdGVzdC5jb20ifQ.UibZSs0Bc0HMaacj7EnGR95X38DH5cTw5Hf90aiUq-y-38xjMIgJYiowa_IM8BujLvB1fHj6ucVniv7yx22uFw'
    expect(store.getters.job).toBe('7')
  })
  it('assumes contractor role with no token', () => {
    store.state.token = ''
    expect(store.getters.isContractorUser).toBe(true)
    store.state.token = undefined
    expect(store.getters.isContractorUser).toBe(true)
  })
  it('recognizes contractor token', () => {
    store.state.token = 'eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJleHAiOjEyMzQ1Njc3LCJyb2xlIjoiY29udHJhY3RvciIsImpvYiI6IjciLCJlbWFpbCI6InRlc3RAdGVzdC5jb20ifQ.cJSBk6Sxiwf194U0sa6XfWfGcpi6G_GNr-AXjl0CQQOZlCfwH254zo7fLTt80tQQgoC5O9JR2K1Hv3et1zzomA'
    expect(store.getters.isContractorUser).toBe(true)
  })
  it('recognizes subcontractor token', () => {
    store.state.token = 'eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJleHAiOjEyMzQ1Njc3LCJyb2xlIjoic3ViY29udHJhY3RvciIsImpvYiI6IjciLCJlbWFpbCI6InRlc3RAdGVzdC5jb20ifQ.UibZSs0Bc0HMaacj7EnGR95X38DH5cTw5Hf90aiUq-y-38xjMIgJYiowa_IM8BujLvB1fHj6ucVniv7yx22uFw'
    expect(store.getters.isContractorUser).toBe(false)
  })
  it('gets exp', () => {
    store.state.token = 'eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJleHAiOjEyMzQ1Njc3LCJyb2xlIjoic3ViY29udHJhY3RvciIsImpvYiI6IjciLCJlbWFpbCI6InRlc3RAdGVzdC5jb20ifQ.UibZSs0Bc0HMaacj7EnGR95X38DH5cTw5Hf90aiUq-y-38xjMIgJYiowa_IM8BujLvB1fHj6ucVniv7yx22uFw'
    expect(store.getters.exp).toBe(12345677)
  })
  it('handles failed refresh', () => {
    const pause = function () {
      // nothing
    }
    store.replaceState({
      status: 'abc',
      token: 'abc',
      user: 'abc',
      refresher: setInterval(pause, 500)
    })
    store.dispatch('refresh')
      .then(() => {
        expect(1).toBe(0) // promise should be rejected
      })
      .catch(() => {
        // ignore the error and do the asserts
      })
    mockAxios.mockError()
    expect(mockAxios.get).toBeCalledWith('/token-refresh')
    expect(store.state.status).toBe('error')
    expect(store.state.token).toBe('')
    expect(store.state.user).toBe('')
    expect(store.state.refresher).toBe('')
  })
  it('handles successful refresh', () => {
    axios.defaults = {}
    axios.defaults.headers = {}
    axios.defaults.headers.common = []
    const pause = function () {
      // nothing
    }
    store.replaceState({
      status: 'abc',
      token: 'abc',
      user: 'abc',
      refresher: setInterval(pause, 500)
    })
    const refresherVal = store.state.refresher
    store.dispatch('refresh')
      .then(() => {
        // nothing
      })
      .catch(() => {
        expect(1).toBe(0) // promise should be resolved
      })
    mockAxios.mockResponse({
      data: {
        token: 'test token'
      }
    })
    expect(mockAxios.get).toBeCalledWith('/token-refresh')
    expect(store.state.status).toBe('success')
    expect(store.state.token).toBe('test token')
    expect(store.state.user).toBe('abc')
    expect(store.state.refresher).toBe(refresherVal)
    expect(axios.defaults.headers.common['Planroom-Authorization']).toBe('Bearer test token')
  })
  it('logs out', () => {
    axios.defaults = {}
    axios.defaults.headers = {}
    axios.defaults.headers.common = []
    axios.defaults.headers.common['Planroom-Authorization'] = 'some token'
    const pause = function () {
      // nothing
    }
    store.replaceState({
      status: 'abc',
      token: 'abc',
      user: 'abc',
      refresher: setInterval(pause, 500)
    })
    store.dispatch('logout')
      .then(() => {
        // nothing
      })
      .catch(() => {
        expect(1).toBe(0) // promise should be resolved
      })

    expect(store.state.status).toBe('')
    expect(store.state.token).toBe('')
    expect(store.state.user).toBe('')
    expect(store.state.refresher).toBe('')
    expect(axios.defaults.headers.common['Planroom-Authorization']).toBe(undefined)
  })
  it('handles failed logins', () => {
    const pause = function () {
      // nothing
    }
    store.replaceState({
      status: 'abc',
      token: 'abc',
      user: 'abc',
      refresher: setInterval(pause, 500)
    })
    store.dispatch('login', {email: 'joe@joe.com', password: 'password'})
      .then(() => {
        expect(1).toBe(0) // promise should be rejected
      })
      .catch(() => {
        // ignore the error and do the asserts
      })
    mockAxios.mockError()
    expect(mockAxios.post).toBeCalledWith('/login', { email: 'joe@joe.com', password: 'password' })
    expect(store.state.status).toBe('error')
    expect(store.state.token).toBe('')
    expect(store.state.user).toBe('')
    expect(store.state.refresher).toBe('')
  })
  it('handles successful logins', () => {
    const pause = function () {
      // nothing
    }
    axios.defaults = {}
    axios.defaults.headers = {}
    axios.defaults.headers.common = []
    store.replaceState({
      status: 'abc',
      token: 'abc',
      user: 'abc',
      refresher: setInterval(pause, 500)
    })
    store.dispatch('login', { email: 'joe@joe.com', password: 'password' })
      .then(() => {
        // ignoring this
      })
      .catch(() => {
        expect(1).toBe(0) // promise should be resolved
      })
    const token = 'eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJleHAiOjk5OTk5OTk5OTksImVtYWlsIjoiSk9FQGpvZS5jb20ifQ.5LAFiI3QdcttM-Eck2sa-7wN2KubmVGhsFOUitG5yg3bUWXNqLfpbX1Tq7Bj9h9efrYJJhCbpjPMowJdM-lHNQ'
    mockAxios.mockResponse({
      data: {
        token
      }
    })
    expect(mockAxios.post).toBeCalledWith('/login', { email: 'joe@joe.com', password: 'password' })
    expect(store.state.status).toBe('success')
    expect(store.state.token).toBe(token)
    expect(store.state.user).toBe('JOE@joe.com') // caps because that's how the token has it
    expect(store.state.refresher).toBeTruthy()
    expect(axios.defaults.headers.common['Planroom-Authorization']).toBe('Bearer ' + token)
  })
  it('stores subcontractor token', () => {
    store.replaceState({
      status: 'abc',
      token: 'abc',
      user: 'abc',
      refresher: 'fakeFunction'
    })

    axios.defaults = {}
    axios.defaults.headers = {}
    axios.defaults.headers.common = []

    const token = 'eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJleHAiOjEyMzQ1Njc3LCJyb2xlIjoiY29udHJhY3RvciIsImpvYiI6IjciLCJlbWFpbCI6InRlc3RAdGVzdC5jb20ifQ.cJSBk6Sxiwf194U0sa6XfWfGcpi6G_GNr-AXjl0CQQOZlCfwH254zo7fLTt80tQQgoC5O9JR2K1Hv3et1zzomA'
    store.dispatch('subcontractorToken', token)

    expect(store.state.status).toBe('manual')
    expect(store.state.token).toBe(token)
    expect(store.state.user).toBe('test@test.com') // caps because that's how the token has it
    expect(store.state.refresher).toBe('')
    expect(axios.defaults.headers.common['Planroom-Authorization']).toBe('Bearer ' + token)
  })
})


src/main.js
// https://github.cm/mjsmith11/lkf
import Vue from 'vue'
import App from './App.vue'
import router from './router'
import store from './store'
import Vuelidate from 'vuelidate'
import axios from 'axios'
import 'bootstrap/dist/css/bootstrap.min.css'
import { library } from '@fortawesome/fontawesome-svg-core'
import { faMapMarkerAlt, faPhone, faFax, faTrashAlt } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'

library.add(faMapMarkerAlt)
library.add(faPhone)
library.add(faFax)
library.add(faTrashAlt)

Vue.component('font-awesome-icon', FontAwesomeIcon)

Vue.config.productionTip = false

Vue.use(Vuelidate)

axios.defaults.baseURL = process.env.VUE_APP_API_URL
axios.defaults.headers.post['Content-Type'] = 'application/json'
axios.defaults.headers.put['Content-Type'] = 'application/json'
axios.defaults.headers.common['Accept'] = 'application/json'

new Vue({
  router,
  store,
  render: h => h(App)
}).$mount('#app')


src/router.js
import Vue from 'vue'
import Router from 'vue-router'
import store from './store'

const Login = resolve => {
  require.ensure(['./views/Login.vue'], () => {
    resolve(require('./views/Login.vue'))
  }, 'contractor')
}

const JobsIndex = resolve => {
  require.ensure(['./views/jobs/JobsIndex.vue'], () => {
    resolve(require('./views/jobs/JobsIndex.vue'))
  }, 'contractor')
}

const NewJob = resolve => {
  require.ensure(['./views/jobs/NewJob.vue'], () => {
    resolve(require('./views/jobs/NewJob.vue'))
  }, 'contractor')
}

const ShowJob = resolve => {
  require.ensure(['./views/jobs/ShowJob.vue'], () => {
    resolve(require('./views/jobs/ShowJob.vue'))
  }, 'show_job')
}

const Invite = resolve => {
  require.ensure(['./views/jobs/Invite.vue'], () => {
    resolve(require('./views/jobs/Invite.vue'))
  }, 'contractor')
}

const ExpiredToken = resolve => {
  require.ensure(['./views/subcontractor/ExpiredToken.vue'], () => {
    resolve(require('./views/subcontractor/ExpiredToken.vue'))
  }, 'expired')
}

Vue.use(Router)

let router = new Router({
  mode: 'history',
  base: process.env.BASE_URL,
  routes: [
    {
      path: '/',
      redirect: '/jobs'
    },
    {
      path: '/jobs',
      name: 'jobs index',
      component: JobsIndex,
      meta: {
        requiresAuth: true
      }
    },
    {
      path: '/jobs/new',
      name: 'new job',
      component: NewJob,
      meta: {
        requiresAuth: true
      }
    },
    {
      path: '/jobs/:id',
      name: 'show job',
      component: ShowJob,
      meta: {
        requiresAuth: true
      }
    },
    {
      path: '/jobs/:id/invite',
      name: 'job invite',
      component: Invite,
      meta: {
        requiresAuth: true
      }
    },
    {
      path: '/expired',
      name: 'expired token',
      component: ExpiredToken
    },
    {
      path: '/login',
      name: 'login',
      component: Login,
      beforeEnter: (to, from, next) => beforeLoginEnter(to, from, next)
    }
  ]
})

export function beforeLoginEnter (to, from, next) {
  if (store.getters.isLoggedIn) {
    next('/jobs')
  } else {
    next()
  }
}

export function beforeEach (to, from, next) {
  if (to.query !== undefined && to.query.token !== undefined) {
    store.dispatch('subcontractorToken', to.query.token)
  }
  if (to.matched.some(record => record.meta.requiresAuth)) {
    if (store.getters.isLoggedIn) {
      if (store.getters.isContractorUser) {
        // logged in contractor
        next()
      } else {
        if (to.path !== '/jobs/' + store.getters.job) {
          // logged in subcontractor who has wrong path
          next('/jobs/' + store.getters.job)
        } else {
          // logged in subcontractor with correct path
          next()
        }
      }
      return
    }
    if (store.getters.isContractorUser) {
      // Contractor user not logged in
      next('/login')
    } else {
      // Subcontractor not logged in
      next('/expired')
    }
  } else {
    // route doesn't require authentication
    next()
  }
}

router.beforeEach((to, from, next) => beforeEach(to, from, next))

export default router


src/mixins/datetimeFilters.js
const helpers = {
  date (value) {
    var splits = value.split('-')
    var dd = splits[2]
    var mm = splits[1]
    var yyyy = splits[0]

    return mm + '/' + dd + '/' + yyyy
  },
  time (value) {
    var AMPM = 'AM'
    var pieces = value.split(':')
    var HH = pieces[0]
    var mm = pieces[1]
    var hh = ''
    if (HH === '00') {
      hh = 12
    } else if (HH > 12) {
      hh = HH - 12
      if (hh < 10) {
        hh = '0' + hh
      }
      AMPM = 'PM'
    } else {
      if (HH === '12') {
        AMPM = 'PM'
      }
      hh = HH
    }
    return hh + ':' + mm + ' ' + AMPM
  },
  isDST (date) {
    // A free script from: www.mresoftware.com
    var yr = date.getFullYear()
    var dstStart = new Date('March 14, ' + yr + ' 02:00:00') // 2nd Sunday in March can't occur after the 14th
    var dstEnd = new Date('November 07, ' + yr + ' 02:00:00') // 1st Sunday in November can't occur after the 7th
    var day = dstStart.getDay() // day of week of 14th
    dstStart.setDate(14 - day) // Calculate 2nd Sunday in March of this year
    day = dstEnd.getDay() // day of the week of 7th
    dstEnd.setDate(7 - day) // Calculate first Sunday in November of this year
    if (date >= dstStart && date < dstEnd) { // does today fall inside of DST period?
      return true // if so then return true
    }
    return false // if not then return false
  }
}
export const datetimeFilters = {
  filters: {
    /**
     *
     * @param { string } value date representation as YYYY-MM-DD
     * @returns { string } date formatted as MM/DD/YYYY
     */
    date (value) {
      if (value === undefined) {
        return ''
      }

      return helpers.date(value)
    },
    /**
     *
     * @param { string } value datetime formatted with YYYY-MM-DD HH:mm:ss (24 hour Eastern time)
     * @returns { string } formatted date time MM/DD/YYYY hh:mm A (12 hour time) with EST or EDT appended appropriately
     */
    datetime (value) {
      if (value === undefined || value === '') {
        return ''
      }
      var pieces = value.split(' ')
      var datePart = helpers.date(pieces[0])
      var timePart = helpers.time(pieces[1])
      var datetime = datePart + ' ' + timePart
      var tz = helpers.isDST(new Date(datetime)) ? ' EDT' : ' EST'
      return datetime + tz
    },
    /**
     *
     * @param { number } value unix timestamp
     * @returns { string } formatted date time YYYY-MM-DD HH:mm:ss (24 hour Eastern time)
     */
    unixDatetime (value) {
      if (value === undefined) {
        return ''
      }

      let date = new Date(value * 1000)
      let minutes = date.getMinutes()
      if (minutes < 10) {
        minutes = '0' + minutes
      }
      return date.getFullYear() + '-' + (date.getMonth() + 1) + '-' + date.getDate() + ' ' + date.getHours() + ':' + minutes + ':00'
    },
    /**
     *
     * @param { string } value datetime formatted with YYYY-MM-DD HH:mm:ss (24 hour Eastern time)
     * @returns { string } formatted date time MM/DD/YYYY hh:mm A (12 hour time adjusted by 1 hour) with CST or CDT appended appropriately
     */
    datetimeCentral (value) {
      if (value === undefined) {
        return ''
      }
      var pieces = value.split(' ')

      var datePiece = pieces[0]
      var timePiece = pieces[1]

      var dateSplit = datePiece.split('-')
      var timeSplit = timePiece.split(':')
      var utc = Date.UTC(dateSplit[0], dateSplit[1] - 1, dateSplit[2], timeSplit[0], timeSplit[1], timeSplit[2]) // months are 0 based

      var d = new Date(utc)
      d.setHours(d.getHours() - 1)

      var mm = d.getMonth() + 1 // months are zero based
      var dd = d.getDate()
      if (mm < 10) {
        mm = '0' + mm
      }
      if (dd < 10) {
        dd = '0' + dd
      }
      var dateString = d.getFullYear() + '-' + mm + '-' + dd
      var hh = d.getUTCHours()
      var min = d.getMinutes()
      var ss = d.getSeconds()
      if (hh < 10) {
        hh = '0' + hh
      }
      if (min < 10) {
        min = '0' + min
      }
      if (ss < 10) {
        ss = '0' + ss
      }
      var timeString = hh + ':' + min + ':' + ss

      var datePart = helpers.date(dateString)
      var timePart = helpers.time(timeString)

      var datetime = datePart + ' ' + timePart
      var tz = helpers.isDST(new Date(datetime)) ? ' CDT' : ' CST'
      return datetime + tz
    }
  }
}


src/store.js
import Vue from 'vue'
import Vuex from 'vuex'
import axios from 'axios'

Vue.use(Vuex)

export default new Vuex.Store({
  state: {
    status: '',
    token: '',
    user: '',
    refresher: ''
  },
  mutations: {
    auth_request (state) {
      state.status = 'loading'
    },
    auth_success (state, data) {
      state.status = 'success'
      state.token = data.token
      state.user = data.user
      state.refresher = data.refresher
    },
    auth_error (state) {
      state.status = 'error'
      state.token = ''
      state.user = ''
      clearInterval(state.refresher)
      state.refresher = ''
    },
    auth_refresh (state, token) {
      state.status = 'success'
      state.token = token
    },
    logout (state) {
      state.status = ''
      state.token = ''
      state.user = ''
      clearInterval(state.refresher)
      state.refresher = ''
    },
    set_token (state, token) {
      state.status = 'manual'
      state.token = token
      const decoded = parseJwt(token)
      state.user = decoded.email
      clearInterval(state.refresher)
      state.refresher = ''
    }
  },
  actions: {
    login ({ commit, dispatch }, user) {
      return new Promise((resolve, reject) => {
        commit('auth_request')
        axios.post('/login', user)
          .then(resp => {
            const token = resp.data.token
            const decoded = parseJwt(token)
            const user = decoded.email
            const refresher = setInterval(dispatch.bind(this, 'refresh'), 14 * 60 * 1000) // every 14 minutes
            axios.defaults.headers.common['Planroom-Authorization'] = 'Bearer ' + token
            commit('auth_success', { token, user, refresher })
            resolve(resp)
          })
          .catch(err => {
            commit('auth_error')
            reject(err)
          })
      })
    },
    logout ({ commit }) {
      return new Promise((resolve, reject) => {
        commit('logout')
        delete axios.defaults.headers.common['Planroom-Authorization']
        resolve()
      })
    },
    refresh ({ commit }) {
      return new Promise((resolve, reject) => {
        commit('auth_request')
        axios.get('/token-refresh')
          .then(resp => {
            const token = resp.data.token
            axios.defaults.headers.common['Planroom-Authorization'] = 'Bearer ' + token
            commit('auth_refresh', token)
            resolve(resp)
          })
          .catch(err => {
            commit('auth_error')
            reject(err)
          })
      })
    },
    subcontractorToken ({ commit }, token) {
      axios.defaults.headers.common['Planroom-Authorization'] = 'Bearer ' + token
      commit('set_token', token)
    }
  },
  getters: {
    isLoggedIn: state => {
      if (!state.token) {
        // if no token, we aren't logged in
        return false
      } else {
        // see if the token is expired
        let decoded = parseJwt(state.token)
        let now = Math.round((new Date()).getTime() / 1000)
        // using a 5 second threshold to give caller time to do something if this is true
        return (decoded.exp - now >= 5)
      }
    },
    authStatus: state => state.status,
    token: state => state.token,
    // assume contractor with no token because a subcontractor should always have one.
    isContractorUser: state => {
      if (state.token === undefined || state.token === '') {
        return true
      } else {
        let decoded = parseJwt(state.token)
        return decoded.role === 'contractor'
      }
    },
    job: state => {
      let decoded = parseJwt(state.token)
      return decoded.job
    },
    exp: state => {
      let decoded = parseJwt(state.token)
      return decoded.exp
    }
  }
})

function parseJwt (token) {
  var base64Url = token.split('.')[1]
  var base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/')
  return JSON.parse(window.atob(base64))
};
